packages/next/src/client/flight-data-helpers.ts

```diff
@@ -0,0 +1,59 @@
+import type {
+  CacheNodeSeedData,
+  FlightDataPath,
+  FlightRouterState,
+  FlightSegmentPath,
+  Segment,
+} from '../server/app-render/types'
+
+export function getFlightDataPartsFromPath(flightDataPath: FlightDataPath): {
+  /**
+   * The full `FlightSegmentPath` inclusive of the final `Segment`
+   */
+  segmentPath: FlightSegmentPath
+  /**
+   * The `FlightSegmentPath` exclusive of the final `Segment`
+   */
+  pathToSegment: FlightSegmentPath
+  segment: Segment
+  tree: FlightRouterState
+  seedData: CacheNodeSeedData | null
+  head: React.ReactNode | null
+} {
+  // tree, seedData, and head are *always* the last three items in the `FlightDataPath`.
+  const [tree, seedData, head] = flightDataPath.slice(-3)
```

Comment:
If it becomes perf critical, this should just access the relevant element via `flightDataPath[flightDataPath.length - N]` to avoid another intermediate array + iterator protocol invocation to destructure (though that may be optimized in runtimes nowadays).